src/Npgsql/Replication/NpgsqlReplicationConnection.cs

```diff
@@ -0,0 +1,888 @@
+
+                    yield return data;
+
+                    // Our consumer may have disposed the stream which isn't necessary but shouldn't hurt us
+                    if (columnStream.IsDisposed)
+                        columnStream = new NpgsqlReadBuffer.ColumnStream(buf);
+                    // Our consumer may not have read the stream to the end, but it might as well have been us
+                    // ourselves bypassing the stream and reading directly from the buffer in StartReplication()
+                    else if (columnStream.Position < columnStream.Length && !bypassingStream)
+                    {
+                        await buf.Skip(columnStream.Length - columnStream.Position, true, CancellationToken.None);
+                    }
+
+                    continue;
+                }
+
+                case 'k': // Primary keepalive message
+                {
+                    await buf.Ensure(17, true, CancellationToken.None);
+                    var endLsn = buf.ReadUInt64();
+                    var timestamp = buf.ReadInt64();
+                    var replyRequested = buf.ReadByte() == 1;
+                    if (unchecked((ulong)Interlocked.Read(ref _lastReceivedLsn)) < endLsn)
+                        Interlocked.Exchange(ref _lastReceivedLsn, unchecked((long)endLsn));
+
+                    if (replyRequested && await _feedbackSemaphore.WaitAsync(Timeout.Infinite, CancellationToken.None))
+                        await SendFeedback(cancellationToken: CancellationToken.None);
+
+                    continue;
+                }
+
+                default:
+                    throw connector.Break(new NpgsqlException($"Unknown replication message code '{code}'"));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Sends a forced status update to PostgreSQL with the current WAL tracking information.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The connection currently isn't streaming</exception>
+        /// <returns>A Task representing the sending of the status update (and not any PostgreSQL response).</returns>
+        public Task SendStatusUpdate(CancellationToken cancellationToken = default)
+        {
+            using (NoSynchronizationContextScope.Enter())
+                return SendStatusUpdateInternal(cancellationToken);
+        }
+
+        async Task SendStatusUpdateInternal(CancellationToken cancellationToken = default)
+        {
+            EnsureState(ReplicationConnectionState.Streaming);
+            if (await _feedbackSemaphore.WaitAsync(Timeout.Infinite, cancellationToken))
+                await SendFeedback(cancellationToken: cancellationToken);
+        }
+
+        async Task SendFeedback(bool requestReply = false, CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                // Disable the timers while we are sending
```

Comment:
But we could be racing with a the timer right now, right? We should probably take _feedbackSemaphore here instead of outside (i.e. in TimerSendFeedback, when receiving a keepalive...)

Reply:
Fully agree.

Reply:
Which method could be racing here?
All callers should be blocked out via `_feedbackSemaphore` here.

Reply:
I think you're right - I misread the code: TimerSendFeedback (and TimerRequestFeedback) acquire the semaphore, so it's impossible for them to be running at this point.

The code looks safe to me (maybe @vonzshik sees something else), but I'd consider refactoring to simplify. Instead of everyone locking outside, I'd move the lock inside and pass a flag saying whether to wait on the semaphore or exit immediately if it could not be taken. This way you can always call SendFeedback, and on a best-effort basis it will sometimes abort and do nothing (because it's already being called somewhere else).

Reply:
I personally do not like having one method taking a lock, and another releasing it. It adds more unnecessary complexity and error prone.

Reply:
> I think you're right - I misread the code

That's really relieving 😅. I was already getting a bit frustrated about my inability to get any concurrency right.

> I'd consider refactoring to simplify. Instead of everyone locking outside, I'd move the lock inside and pass a flag saying whether to wait on the semaphore or exit immediately if it could not be taken.

If you think that simplifies it...

You just made a good case that it's a bit complex as it stands.

Reply:
Refactored to handle all taking/releasing of the feedback semaphore inside SendFeedback: 15aec30434a43d3387942d16bcd732835349784a

Reply:
> I personally do not like having one method taking a lock, and another releasing it. It adds more of unnecessary complexity and error prone.

I very much agree :) Concurrency and locking are already complicated enough :rofl: